### PR TITLE
pkg/node: omit the generated spec.ipam.podCIDRs when running with some IPAM modes

### LINKED ISSUE #9409: ENI: Do not fill ciliumnodes with allocated podCIDRs in ENI mode
In ENI mode, `spec.ipam.podCIDRs` of `ciliumnodes` is filled with CIDR allocated by cilium and it's never used. Remove `spec.ipam.podCIDRs` to avoid confusion with ENI IP addresses allocated by operator.

### PR DESCRIPTION
**Pull request by mvisonneau**
Currently, when using Cilium with IPAM in AWS ENI, Azure & AlibabaCloud methods. A random podCIDR block is being generated when the CiliumNode is created although is not being leveraged.

```yaml
[...]
spec:
  ipam:
    podCIDRs:
    - 10.194.0.0/16 # << this one
    pool:
      172.16.5.192:
        resource: eni-0f2cda9c5fe02382b
      172.16.5.193:
        resource: eni-0f2cda9c5fe02382b
      172.16.5.194:
        resource: eni-0f2cda9c5fe02382b
```

To avoid operators confusion and make it easier to troubleshoot, I felt it could be interesting propose this change which omits the generation of this CIDR block.

```yaml
[...]
spec:
  ipam:
    pool:
      172.16.5.192:
        resource: eni-0f2cda9c5fe02382b
      172.16.5.193:
        resource: eni-0f2cda9c5fe02382b
      172.16.5.194:
        resource: eni-0f2cda9c5fe02382b
```

Fixes https://github.com/cilium/cilium/issues/9409

### CODE REVIEW

File: pkg/node/address.go [hunk 1]
```diff
@@ -7,6 +7,7 @@ import (
 	"bufio"
 	"errors"
 	"fmt"
+	ipamOption "github.com/cilium/cilium/pkg/ipam/option"
```
Comment (thorn3r) [inline pkg/node/address.go]: you can run `goimports -w -local github.com/cilium/cilium address.go` which will reorder the imports to resolve this linting error

**Comment by thorn3r** [comment]
thanks for the PR @mvisonneau! looks like the checkpatch test is failing because the commit title is too long. Mind amending that to shorten it a bit? I'll kick off the rest of the test suite

File: pkg/node/address.go (hunk 1 above)
Comment (mvisonneau) [reply pkg/node/address.go]: ah indeed my bad, thanks! 🙇

**Comment by mvisonneau** [comment]
Thanks for the review @thorn3r! I updated the commit message.

**Review by thorn3r — APPROVED** [review]
LGTM! seems like we hit an already tracked flake

File: pkg/node/address.go [hunk 2]
```diff
@@ -423,6 +425,11 @@ func AutoComplete() error {
 
 	InitDefaultPrefix(option.Config.DirectRoutingDevice)
 
+	// in ENI mode, we do not leverage these generated CIDRs
+	if option.Config.IPAM == ipamOption.IPAMENI {
```
Comment (jaffcheng) [inline pkg/node/address.go]: Should we include Azure and Alibabacloud here as well?

**Review by jaffcheng — COMMENTED** [review]
Thank you! I think this should close https://github.com/cilium/cilium/issues/9409

File: pkg/node/address.go (hunk 2 above)
Comment (mvisonneau) [reply pkg/node/address.go]: definitely, I won't be able to test **AlibabaCloud** but I updated this change to cover all 3!

**Comment by jaffcheng** [comment]
Code changes LGTM! I'm not familiar with cilium CI workflows, looks like the checkpatch test fails with the current commit title, and maybe we should remove the second commit.
> Error: ERROR:CUSTOM: Please avoid long commit subjects (max: 75, found: 81)

**Comment by mvisonneau** [comment]
👋 thanks for the review @jaffcheng. Not sure how this merge commit ended up there, I removed it and also shorten the message of the relevant one!

**Review by jaffcheng — APPROVED** [review]

**Comment by aanm** [comment]
@christarazi FYI

File: pkg/node/address.go [hunk 3]
```diff
@@ -423,6 +425,12 @@ func AutoComplete() error {
 
 	InitDefaultPrefix(option.Config.DirectRoutingDevice)
```
Comment (jaffcheng) [inline pkg/node/address.go:426]: @mvisonneau  I'm a bit lost about how this work, could you share more on that? I assume the cidr is generated in `InitDefaultPrefix`, but it's not skipped.
https://github.com/cilium/cilium/pull/23048/files#diff-9143180298a1f5ba5dc10fd98d87945d1381ef45b7347a639f85e13e84285ca2L424

Comment (mvisonneau) [reply pkg/node/address.go:426]: absolutely, I did not find it useful to skip the `InitDefaultPrefix` function call during my tests. I assumed that the allocation was happening later ([L633](https://github.com/cilium/cilium/pull/23048/files#diff-9143180298a1f5ba5dc10fd98d87945d1381ef45b7347a639f85e13e84285ca2R633) &/or [L644](https://github.com/cilium/cilium/pull/23048/files#diff-9143180298a1f5ba5dc10fd98d87945d1381ef45b7347a639f85e13e84285ca2R644)).

As we return early ([L431](https://github.com/cilium/cilium/pull/23048/files#diff-9143180298a1f5ba5dc10fd98d87945d1381ef45b7347a639f85e13e84285ca2R431)) it seems to be sufficient. Not sure about the impact of moving the return logic before the  `InitDefaultPrefix` call but happy to try it out if you want/prefer.

Comment (jaffcheng) [reply pkg/node/address.go:426]: Thanks for clarifying.
Looks like there are two places calling `InitDefaultPrefix`:
https://github.com/cilium/cilium/blob/5ff3d72c8b52be779d7bfab8b6e6abfc9f3b8137/pkg/node/address.go#L426  
https://github.com/cilium/cilium/blob/5ff3d72c8b52be779d7bfab8b6e6abfc9f3b8137/daemon/cmd/daemon_main.go#L1404  
I tried the patch in Alibabacloud mode and found that I have to skip both `InitDefaultPrefix` calls to omit the podCIDR. Not sure if there are some differences between ENI mode.

File: pkg/node/address.go [hunk 4]
```diff
@@ -517,7 +525,7 @@ func chooseHostIPsToRestore(ipv6 bool, fromK8s, fromFS net.IP, cidrs []*cidr.CID
 	}
 
 	for _, cidr := range cidrs {
-		if cidr != nil && cidr.Contains(ip) {
+		if cidr != nil && cidr.String() != "" && cidr.Contains(ip) {
```
Comment (christarazi) [inline pkg/node/address.go:528]: Why is this change necessary? Shouldn't `cidr != nil` result in the same thing as `cidr.String() != ""`?

**Review by christarazi — CHANGES_REQUESTED** [review]
Thanks for the PR. I don't quite understand why any of these changes are needed. Shouldn't it be as simple as detecting which IPAM mode we need to skip updating the podCIDR field for during the CiliumNode creation? I'd expect such a change to be [here](https://github.com/cilium/cilium/blob/5abb0144c0024140f6aa812913b7c47e8ae3fe0b/pkg/nodediscovery/nodediscovery.go#L499-L500).

**Comment by christarazi** [comment]
@mvisonneau Do you plan to continue on this PR?

**Comment by mvisonneau** [comment]
I was yes, I apologize as I haven't been able to prioritize it yet. Feel free to update or make progress onto if necessary though!

**Comment by aanm** [comment]
@mvisonneau ping? :slightly_smiling_face:

**Comment by joestringer** [comment]
Thanks for the contribution. I've marked the PR as draft while the feedback is addressed. When the PR is ready for review, you can indicate this by clicking the "Ready for review" button at the bottom of the page.

**Comment by mvisonneau** [comment]
My apologies, I am going to put my head back onto it and will keep you posted!

**Comment by gandro** [comment]
I have pushed a PR that implements Chris' suggestion to not announce the podCIDR for now: https://github.com/cilium/cilium/pull/26663

Long term however, I think we do want to fully get rid of the code that auto-generates the podCIDR for _all IPAM modes_. It's not used by any IPAM mode, so let's just remove that.

**Comment by christarazi** [comment]
@mvisonneau Any update?

**Comment by jaffcheng** [comment]
Looks like the related issue is fixed by https://github.com/cilium/cilium/pull/26663 ?

**Comment by gandro** [comment]
> Looks like the related issue is fixed by #26663 ?

Yes, that PR fixed the issue for ENI. There is still more work around the use of pod CIDRs in IPAM modes that don't have them (c.f. https://github.com/cilium/cilium/issues/27018), but the issue addressed by this PR should be fixed now.

@mvisonneau I'm closing this PR as "obsolete". Please feel free to re-open or ping me if you feel otherwise.

**Comment by mvisonneau** [comment]
Apologies, I had messed up with my email notifications! 🙈 Amazing, thanks, I'll try it out but at first sight it does indeed seem to supersedes it 👍
